includes/class-wc-payment-tokens.php

```diff
@@ -0,0 +1,250 @@
+		) );
+		foreach ( $args as $meta_key => $meta_value ) {
+			if ( in_array( $meta_key, $existing_meta ) ) {
+				$wpdb->update( $wpdb->prefix . 'woocommerce_payment_token_meta', array(
+					'meta_value' => $meta_value
+				), array( 'token_id'   => $token_id, 'meta_key'   => $meta_key ) );
+			} else {
+				$wpdb->insert( $wpdb->prefix . 'woocommerce_payment_token_meta', array(
+					'meta_key'   => $meta_key,
+					'meta_value' => $meta_value,
+					'token_id'   => $token_id
+				) );
+			}
+		}
+
+		do_action( 'woocommerce_payment_token_updated', $token_id );
+		return self::generate_token( $token_id, $core_fields_to_update );
+	}
+
+	/**
+	 * Remove a payment token from the database
+	 * @param int $token_id Token ID
+	 */
+	public static function delete( $token_id ) {
+		global $wpdb;
+		$wpdb->delete( $wpdb->prefix . 'woocommerce_payment_tokens', array( 'token_id' => $token_id ), array( '%d' ) );
+		$wpdb->delete( $wpdb->prefix . 'woocommerce_payment_token_meta', array( 'token_id' => $token_id ), array( '%d' ) );
+		do_action( 'woocommerce_payment_token_deleted', $token_id );
+	}
+
+	/**
+	 * Returns an array of payment token objects associated with the passed customer ID
+	 * @param int $customer_id Customer ID
+	 * @return array Array of token objects
+	 */
+	public static function get_customer_tokens( $customer_id ) {
+		if ( $customer_id < 1 ) {
+			return array();
+		}
+
+		global $wpdb;
+
+		$token_results = $wpdb->get_results( $wpdb->prepare(
+			"SELECT * FROM {$wpdb->prefix}woocommerce_payment_tokens WHERE customer_id = %d",
+			$customer_id
+		) );
+
+		if ( empty( $token_results ) ) {
+			return array();
+		}
+
+		$tokens = array();
+		foreach ( $token_results as $token_result ) {
+			$_token = self::generate_token( $token_result->token_id, (array) $token_result );
+			if ( ! empty( $_token ) ) {
+				$tokens[$token_result->token_id] = $_token;
+			}
+		}
+
+		return $tokens;
+	}
+
+	/**
+	 * Returns an array of payment token objects associated with the passed order ID
+	 * @param int $order_id Order ID
+	 * @return array Array of token objects
+	 */
+	public static function get_order_tokens( $order_id ) {
+		$order = wc_get_order( $order_id );
+
+		if ( ! $order ) {
+			return array();
+		}
+
+		$wc_tokens_meta = get_post_meta( $order_id, '_wc_payment_tokens', true );
+
+		if ( empty ( $wc_tokens_meta ) ) {
+			return array();
+		}
+
+		global $wpdb;
+
+		$wc_tokens_meta_string = implode( ',', array_map( 'intval', $wc_tokens_meta ) );
+		$token_results = $wpdb->get_results(
+			"SELECT * FROM {$wpdb->prefix}woocommerce_payment_tokens WHERE token_id IN ( {$wc_tokens_meta_string} )"
+		);
+
+		if ( empty( $token_results ) ) {
+			return array();
+		}
+
+		$tokens = array();
+		foreach ( $token_results as $token_result ) {
+			$_token = self::generate_token( $token_result->token_id, (array) $token_result );
+			if ( ! empty( $_token ) ) {
+				$tokens[$token_result->token_id] = $_token;
```

Comment:
WP coding standards require a [whitespace around the index if it's a variable](https://make.wordpress.org/core/handbook/best-practices/coding-standards/php/#space-usage) (i.e. this should be `$tokens[ $token_result->token_id ]`). :smile: